### PR TITLE
`Exam mode`: Improve exam import validation

### PR DESCRIPTION
**Pull request by laadvo**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

When importing an exam or an exercise group it is possible to input the same short name and/or title for multiple programming exercises. This then leades to issues when creating a project on VSC/CI. In this PR a check was added making sure that titles and short names of all programming exercises being imported are unique.

### Description

When the same short name or title is inputed for multiple programming exercise, an alert is sent informing the instructor of this.  The effected exercise's short name / title is then removed and has to be set again. In case there are no duplicated short names and titles, it is then checked if a project with the same key and name already exists on the VSC/CI as before.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Exam with >1 programming exercise to be imported
- 1 Exercise Group with >1 programming exercise to be imported

1. Go to Course Management and select a course
2. Click on Exams and there the Import an exam button
4. Select an exam with at least 2 programming exercise
5. Use the same title / short name for multiple programming exercises and click on import
6. An Alert should appear and the corresponding duplicated field should now be empty and marked with red
7. Go to Exercise Groups and select Import Exercise Group
8. Select an an exercise group with at least 2 programming exercises
9. Again set the same short name, title and import
10. Alert should appear and the duplicated field should now be empty

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

same as develop

### Screenshots

After clicking imput, the duplicated title is removed and has to be set again:
image

https://github.com/ls1intum/Artemis/assets/73339358/5bb45d7a-f574-4b08-b6ce-0de0b66e2e1e

### CODE REVIEW

File: src/main/webapp/i18n/en/exam.json [hunk 1]
```diff
@@ -522,6 +522,8 @@
                     "shortNameTooltip": "The shortname of a programming task must be unique. The short name of the exercise can NOT be edited after creating the exercise! The short name must begin with a letter, must not contain any special characters and must be at least 3 characters long",
                     "invalidKey": "The chosen short name and title of {{ number }} programming exercise(s) already exists. Please insert a new short name and title for the corresponding programming exercise(s)",
                     "invalidExerciseConfiguration": "One or more of the values specified are invalid. Please change them before importing",
+                    "duplicatedProgrammingExerciseShortName": "Multiple programming exercises in the exam have the same short name. Please choose a different short name!",
+                    "duplicatedProgrammingExerciseTitle": "Multiple programming exercises in the exam have the same title. Please choose a different title!",
```
Comment (DominikRemo) [inline src/main/webapp/i18n/en/exam.json:526]: In [src/main/webapp/i18n/de/exam.json](https://github.com/ls1intum/Artemis/pull/6877/files#diff-c782a9beef384a868bd33101e93f6e4d75301dc8dd025b7202704efd8163fb98) these lines where added directly beneath the entry for "invalidKey". It would be better to move those lines between "invalidKey" and "invalidExerciseConfiguration", too.

**Review by DominikRemo — CHANGES_REQUESTED** [review]
Code review with some suggestions

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 2]
```diff
@@ -138,34 +134,69 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
     }
 
     /**
-     * Checks the programming exercises of the given exercise groups for the uniqueness of the projectKey chosen by the user.
-     * If a non-unique project key is discovered, the short name is resettet, so that the user can input a new one
+     * Checks that all programming exercises of the given exercise group have a unique title and short name.
+     * Additionally, checks if an exercise with the same project key or name already exists on the VSC / CI.
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java]: done

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 3]
```diff
@@ -138,34 +134,69 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
...
+
+        if (shortNames.size() != uniqueShortNames.size()) {
+            exerciseGroups.forEach(exerciseGroup -> exerciseGroup.getExercises().forEach(exercise -> {
+                if (!uniqueShortNames.contains(exercise.getShortName())) {
+                    exercise.setShortName("");
+                }
+                else {
+                    uniqueShortNames.remove(exercise.getShortName());
+                }
+            }));
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java]: Here as well, I now first check if the exercise is of type programming

File: src/main/webapp/i18n/en/exam.json (hunk 1 above)
Comment (laadvo) [reply src/main/webapp/i18n/en/exam.json:525]: The lines are now moved :+1:

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 4]
```diff
@@ -138,34 +134,73 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
...
+                        uniqueTitles.remove(exercise.getTitle());
                     }
                 }
-            });
+            }));
+            throw new ExamConfigurationException(exerciseGroups, 0, "duplicatedProgrammingExerciseTitle");
+        }
+
+        // check for duplicated short names
+        List<String> shortNames = exerciseGroups.stream().flatMap(group -> group.getExercises().stream()).filter(ex -> ex instanceof ProgrammingExercise)
```
Comment (rriyaldhi) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:165]: It would be great if the logic to check duplicated titles & names is moved into a method that can be used for both.

File: src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseTestService.java [hunk 5]
```diff
@@ -607,13 +607,11 @@ void createProgrammingExerciseForExam_invalidExercise_dates(InvalidExamExerciseD
     // TEST
     void createProgrammingExerciseForExam_DatesSet() throws Exception {
         setupRepositoryMocks(examExercise, exerciseRepo, solutionRepo, testRepo, auxRepo);
-        ExerciseGroup exerciseGroup = examExercise.getExerciseGroup();
```
Comment (rriyaldhi) [inline src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseTestService.java:610]: Why is this removed?

**Review by rriyaldhi — COMMENTED** [review]
Change looks good, but left some suggestions and question.

File: src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseTestService.java (hunk 5 above)
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/exercise/programmingexercise/ProgrammingExerciseTestService.java:610]: This is due to the chage I made to `generateProgrammingExerciseForExam`. This test is the only one that did not expect an exam programming exercise being added to the exercise group per default. I think if an exam programming exercise is created and we have `exercise.setExerciseGroup(exerciseGroup);` we should also set `exerciseGroup.addExercise(programmingExercise);`. The last assertion in the test was in my option not really needed, since a Bad Request already implies the post call failing and the exercise not being added to the exercise group

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 6]
```diff
@@ -138,34 +134,73 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
...
+                    }
+                    else {
+                        uniqueShortNames.remove(exercise.getShortName());
+                    }
+                }
+            }));
+            throw new ExamConfigurationException(exerciseGroups, 0, "duplicatedProgrammingExerciseShortName");
+        }
+
+        // Flag to determine, if a programming exercise with an invalid shortName was found
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java]: There are some already existing tests that test the behavior in case the project already exists on VCS/CI. I didn't really change anything there. I moved the forEach into a seperate method :+1:

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 7]
```diff
@@ -138,33 +134,93 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
...
+     * @throws ExamConfigurationException in case of duplicated titles / short names or if one or more programming exercise project keys are not unique
+     */
+    private void preCheckProgrammingExercisesForTitleAndShortNameUniqueness(List<ExerciseGroup> exerciseGroups, String courseShortName) {
+        List<Exercise> programmingExercises = exerciseGroups.stream().flatMap(group -> group.getExercises().stream())
+                .filter(exercise -> exercise.getExerciseType() == ExerciseType.PROGRAMMING).toList();
+
+        // check for duplicated titles
+        boolean duplicatedTitles = checkForAndRemoveDuplicatedTitlesAndShortNames(programmingExercises, true);
+        if (duplicatedTitles) {
+            throw new ExamConfigurationException(exerciseGroups, 0, "duplicatedProgrammingExerciseTitle");
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:213]: The number is not really used, this is why I set it to 0. The error message only says "Multiple" instead of the concrete number because I think it is more confusing than informative to give a number, since the user cannot be sure how the duplicates are counted. Number of all duplicates together or how many different duplicates there are? Still I have to give a number to be able to reuse the `ExamConfigurationException` that was already used before the PR

**Review by DominikRemo — APPROVED** [review]
Thank you for implementing my suggestions

**Review by rriyaldhi — APPROVED** [review]
Thanks for applying the suggestions.

**Review by dearjasmina — APPROVED** [review]
tested on ts2, works as described :)

**Review by pal03377 — APPROVED** [review]
Tested on TS2. I got all the described warnings for all combinations of naming conflicts. Everything worked as expected 🎉 

One additional thing I noticed is that you can give different Exercise Groups different names. This may be confusing, but it does not cause problems.

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java [hunk 8]
```diff
@@ -138,33 +134,93 @@ public List<ExerciseGroup> importExerciseGroupsWithExercisesToExistingExam(List<
...
+    /**
+     * Checks if a project with the same key and name already exists on VCS / CI. The number of such occurrences is counted
+     *
+     * @param programmingExercises that are checked for an existing project
+     * @param courseShortName      the short name of the course the exercises will be imported into
+     * @return the number of exercises that need to be renamed in the client
+     */
+    private AtomicInteger checkForExistingProjectAndRemoveTitleShortName(List<Exercise> programmingExercises, String courseShortName) {
+        // Count how many programming exercises have conflicts with VCS / CI due to the project with the same key / name already existing
+        AtomicInteger numberOfInvalidProgrammingExercises = new AtomicInteger();
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:181]: I did not want to go over all elements of the list again. We can increment the atomic integer every time an exercise is invalid. Or do I understand something incorrectly?

File: src/main/webapp/app/exam/manage/exams/exam-exercise-import/exam-exercise-import.component.ts [hunk 9]
```diff
@@ -179,8 +190,10 @@ export class ExamExerciseImportComponent implements OnInit {
      * @param exercise the exercise to be checked
      */
     validateTitleOfProgrammingExercise(exercise: Exercise): boolean {
-        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
-        return exercise.title?.length! > 0 && this.titleNamePattern.test(exercise.title!) && exercise.title !== this.getBlocklistTitleOfProgrammingExercise(exercise.id!);
+        return (
+            // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
+            exercise.title?.length! > 0 && this.titleNamePattern.test(exercise.title!) && exercise.title !== this.getBlocklistTitleOfProgrammingExercise(exercise.id!)
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/webapp/app/exam/manage/exams/exam-exercise-import/exam-exercise-import.component.ts:195]: this does not work for me: `Returned expression type 0 | undefined | false | boolean is not assignable to type boolean `. I think I will just keep it as it is if that is ok

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (laadvo) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java]: done, thanks for the suggestion!

**Review by Strohgelaender — APPROVED** [review]
code looks good, thanks for implementing the change requests.

**Review by DominikRemo — APPROVED** [review]
Reapprove code

**Review by rriyaldhi — APPROVED** [review]

**Review by jakubriegel — APPROVED** [review]
Works on ts1 👍🏻

**Review by Strohgelaender — APPROVED** [review]
Tested on TS1, did not find any issues. Importing wrongly configured exams or exercise groups leads to the error message and allows to re-input the name.

**Review by tobias-lippert — APPROVED** [review]
Tested on ts1. Validation works.
